File: scripts/rle_arithmetic.R
source("scripts/helpers.R")

operation = snakemake@params[["code"]]
operation = paste0("f <- function(c1p, c2p, c1m, c2m){\n", operation, "\nreturn(result)}")

eval(parse(text=operation))
print(operation)

filetype = snakemake@wildcards[["filetype"]]

fc = snakemake@input[["chip"]]
fb = snakemake@input[["background"]]

c = file_to_grange(fc)
b = file_to_grange(fb, filetype)

cp = coverage(c[strand(c) == "+"])
cm = coverage(c[strand(c) == "-"])

bp = coverage(b[strand(b) == "+"])
bm = coverage(b[strand(b) == "-"])


start.time <- Sys.time()
result = f(cp, bp, cm, bm)
end.time <- Sys.time()

time.taken <- end.time - start.time

time.taken <- as.numeric(time.taken, units="secs")

write(time.taken, file=snakemake@output[["time"]])

result_string = capture.output(print(result))
write(result_string, snakemake@output[["result"]])

File: scripts/rle_arithmetic.py

from helpers import read_file, file_to_coverage, init_ray; init_ray(snakemake.wildcards)

from pyrle import PyRles

from time import time
import datetime
import numpy as np

c1 = file_to_coverage(snakemake.input.chip)
c2 = file_to_coverage(snakemake.input.background)

code = snakemake.params.code

# pyop = {"add": "__add__", "subtract": "__sub__", "multiply": "__mul__", "divide": "__truediv__"}[operation]

# m = getattr(chip, pyop)

start = time()

# result = m(background)
exec(code)

end = time()
total = end - start

total_dt = datetime.datetime.fromtimestamp(total)

minutes_seconds = total_dt.strftime('%-M.%-S.%f')

open(snakemake.output[0], "w+").write(minutes_seconds)

result_string = str(result)

open(snakemake.output.result, "w+").write(result_string)
